src/Language/PureScript/TypeChecker.hs

```diff
@@ -173,36 +172,38 @@ addTypeClass _ qualifiedClassName args implies dependencies ds kind = do
       env <- getEnv
       implies' <- (traverse . overConstraintArgs . traverse) replaceAllTypeSynonyms implies
       let ctIsEmpty = null classMembers && all (typeClassIsEmpty . findSuperClass env) implies'
-      pure $ makeTypeClassData args classMembers implies' dependencies ctIsEmpty
+      let (determinedArgs, coveringSets) = computeCoveringSets (length args) dependencies
+      classMembers' <- traverse (computeVtasNeededForMember (S.toList coveringSets) (typeSynonyms env) (types env)) classMembers
+      pure $ TypeClassData args classMembers' implies' dependencies determinedArgs coveringSets ctIsEmpty
       where
       findSuperClass env c = case M.lookup (constraintClass c) (typeClasses env) of
         Just tcd -> tcd
         Nothing -> internalError "Unknown super class in TypeClassDeclaration"
 
-    coveringSets :: TypeClassData -> [S.Set Int]
-    coveringSets = S.toList . typeClassCoveringSets
-
-    argToIndex :: Text -> Maybe Int
-    argToIndex = flip M.lookup $ M.fromList (zipWith ((,) . fst) args [0..])
+    -- Before Visible Type Applications (VTAs) were introduced, if a type class member
+    -- did not reference all of the type variables in the type class head,
+    -- we would throw an UnusableDeclaration error. With the advent of VTAs,
+    -- we can remove this error and instead inform the user that these type class members
+    -- can only have their corresponding type class instance found if one uses VTAs
+    -- for one of the given sets.
+    computeVtasNeededForMember 
+      :: [S.Set Int]
+      -> SynonymMap 
+      -> KindMap 
+      -> (Ident, SourceType) 
+      -> m (Ident, SourceType, Maybe [[Text]])
+    computeVtasNeededForMember coveringSets syns kinds (ident, memberTy) = do
+      memberTy' <- replaceAllTypeSynonymsM syns kinds memberTy
+      let mentionedArgIndexes = S.fromList (mapMaybe argToIndex (freeTypeVariables memberTy'))
+      let leftovers = map (`S.difference` mentionedArgIndexes) coveringSets
+      pure (ident, memberTy, if any null leftovers then Nothing else Just $ map (map (fst . (args !!)) . S.toList) leftovers)
+      where
+        argToIndex :: Text -> Maybe Int
+        argToIndex = flip M.lookup $ M.fromList (zipWith ((,) . fst) args [0..])
```

Comment:
Can this logic live in `makeTypeClassData`, where the covering sets are originally computed? Then you wouldn't need to expose `computeCoveringSets` and run that calculation again above; I think that would also keep you from having to stuff a dummy `Nothing` in the type class data.

You won't have `replaceAllTypeSynonymsM` there, but all you're using `memberTy'` for is determining free variables, and type synonyms can't introduce free variables. (Actually, I don't really understand why you added it *here*, given that, but whatever.)

Reply:
I put it there initially because my first attempt at updating `makeTypeClassData` resulted in a cycle in module dependencies. So, to get this going, I moved it to where the `UnusableDeclaration` was and finished that part. It stayed there because my next attempt at moving it back into `makeTypeClassData` and resolving the module dependency cycle didn't work.

`SynonynMap` and `KindMap` synonyms are defined in `TypeChecker/Synonyms.hs`. `KindMap` references a type defined in `Environment.hs`. So, it seemed I would need to inline at least those parts into `Environment.hs`. However, due to the import of `Errors.hs` for the `MultipleErrors` type, a cyclical dependency appeared.

If we don't need `replaceAllTypeSynonymsM` in the first place, then it should be possible to move it there.

Reply:
Code has been moved.